### ISSUE TITLE
Interest in a second plugin design?

### ISSUE THREAD

**Issue by wireddown** [zeograd/gcodeutils#10]
_Aside: Thanks for [reviewing and merging](#9) the unit test fixes. I was unsure how active this library is :)_

# RFC

I've prototyped a second plugin design that might make adding new ones easier for folks, and I wanted to gauge your interest in accepting another design.

The prototype (and its unit test) is in this [commit](https://github.com/wireddown/gcodeutils/commit/df9907a206a1c7880c03f876181ea47642873ec3?diff=unified), and this [commit](https://github.com/wireddown/gcodeutils/commit/a82e5a2b453ff17ba909efcde781ddc548b9b363?diff=split) shows how it could extend the existing tools.

How useful would something like this be?

## Motivation

The filter classes have two drawbacks:

* The filters build two kinds of state as they pass over gcode, but they only need one:
  * The work they care about (eg, collecting a group of linear segments and converting them to an arc)
  * Their position in the gcode in support of the work they care about. This state can be tracked elsewhere and not repeated by each filter.
* The filters do not differentiate between the two flavors of GCoder layers: layers with or without extrusion
  * When GCoder encounters pure movement commands, usually in the prologue and epilogue stanzas from the slicer, it groups them as a "layer", but this group is not represented in the object's all_zs property.
  * Thus, for GCoder objects, `len(self.all_layers) != len(self.all_zs)`
  * Operations on gcode that require an awareness of the *printed* layer need to track additional state, but this is not the responsibility of a filter.

To address these, the iterator class tracks the visitor's place in the gcode and then passes that information to the visitor at each step so that the visitor only needs to keep state that directly helps its task. This improves the ["signal to boilerplate" ratio](https://github.com/wireddown/gcodeutils/blob/df9907a206a1c7880c03f876181ea47642873ec3/gcodeutils/visit/pause_at_layer.py). Part of the state passed to the visitor is a bool that represents whether or not the layer 'is_printed'.

**Comment by zeograd** [comment]
>  I was unsure how active this library is :)

Me too :)
I don't think there's anyone using it and I've pivoted to other projects as I print less stuff personally. But I welcome PRs and review them as good as possible.

I pondered a lot about the filter API back then and chose a minimalist one expecting to offer more boilerplate/factorised wrapping as needs arise.
You did well to improve the ease of writing filters.

Feel free to add yourself as author in the gcode_mod CLI and in AUTHORS in a PR

Thanks for your interest in this project :)
